### ISSUE TITLE
Request to add a runtime enum detection mechanism

### ISSUE THREAD

**Issue by jaredjj3** [acro5piano/typed-graphqlify#266]
First of all, thank you for maintaining such an excellent library! Clientside GraphQL in my application doesn't feel brittle anymore.

**TL;DR** - It would be nice to be able to detect whether a property is an enum at runtime.

I've created a wrapper around `typed-graphqlify` and `graphql-code-generator` that allows me to make type-safe queries tied to a generated GraphQL schema type map like this:

```ts
// can't use any string, Mutation must have a keyof 'signup'
const signup = Gql.mutation('signup')
  .setQuery({
    // properties must match DeepPartial<Mutation['signup']>
    id: Gql.string,
    email: Gql.string,
    username: Gql.string,
    role: Gql.optional.oneOf(UserRoles)!,
    confirmedAt: Gql.string,
  })
  // the developer is responsible for using the right input
  .setVariables<{ input: SignupInput }>({
    input: {
      email: Gql.string,
      password: Gql.string,
      username: Gql.string,
    },
  })
  .build();
  
console.log(signup.toString({ input: { ... } })); // query{...}
```

The code that achieves this is in [$gql.ts](https://github.com/jaredjj3/stringsync/blob/24ab8a0c7ea964881ff10f8809b4381fe898bf42/web/src/graphql/%24gql.ts)

When a `variable` is an enum, I need to account for it at runtime so it doesn't get quoted by `rawString`. It's impossible to "correctly" distinguish a string enum from a string using inherent properties. However, I made a solution where the caller manually declares object paths to the enums. As I'm grapqhlifying `variables`, I track the path and skip quoting values that match the object paths. The resulting public API is very awkward:

[Path.ts](https://github.com/jaredjj3/stringsync/blob/24ab8a0c7ea964881ff10f8809b4381fe898bf42/web/src/graphql/Path.ts)

```ts
const changeRoles = Gql.mutation('changeRoles')
  .setQuery([{
    id: Gql.string, role: Gql.optional.oneOf(UserRoles)!
  }])
  .setVariables<{ input: ChangeRolesInput }>({
    input: [{ userId: Gql.string, role: Gql.optional.oneOf(UserRoles)! }],
  })
  // in the variables, don't quote any objects matching object path: input[*].role
  .setEnumPaths([Path.create('input', Path.STAR, 'role')])
  .build();
```

This approach works, but it's error-prone and awkward. It would be nice to be able to detect enums at runtime readily. Since enums use the `oneOf` type definition, perhaps that method can attach extra metadata to demarcate enum fields.

**Comment by acro5piano** [comment]
Hi, thanks for your interest in this library.

> would be nice to be able to detect enums at runtime readily

typed-graphqlify doesn't do anything fancy, so runtime checking shouldn't be implemented.

I'm not looking into your code, but it seems you can use variables as string without worrying about quotation.

Otherwise, I guess you are doing over engineering. If you are using graphql-code-generator then just only use it and forget about typed-graphqlify because that's much simpler to use for complex use-case.

**Comment by acro5piano** [comment]
P.S. Your website looks cool!

**Comment by jaredjj3** [comment]
Thanks for the feedback! What you suggested works, but I'm still faced with the same problem of identifying what fields are enums.

> typed-graphqlify doesn't do anything fancy, so runtime checking shouldn't be implemented.

I agree. What I'm really requesting is extra metadata for fields declared with `types.oneOf`.

### Alternative 1

I don't fully understand how [scalarType](https://github.com/acro5piano/typed-graphqlify/blob/4b9d6d2bd1466dc3e503b1220a20abf8f554133c/src/types.ts#L50) is used, but would you consider adding an additional property that marks the value as an enum?

[types.ts](https://github.com/acro5piano/typed-graphqlify/blob/4b9d6d2bd1466dc3e503b1220a20abf8f554133c/src/types.ts)

```ts
export class types {
  static oneOf<T extends ReadonlyArray<string>>(_e: T, meta?: any): ElementType<T>
  static oneOf<T extends {}>(_e: T, meta?: any): ValueOf<T> | keyof T
  static oneOf<T extends {} | ReadonlyArray<string>>(_e: T, meta?: any) {
    return scalarType(meta)
  }
}

function scalarType(meta?: any): any {
  const scalar: GraphQLScalar = {
    [typeSymbol]: GraphQLType.SCALAR,
    meta,
  }
  return scalar
}
```

### Alternative 2

Another possibility is to not use a unique symbol for `typeSymbol`, and allow `typed-graphqlify` adopters to author their own type producers:

[render.ts](https://github.com/acro5piano/typed-graphqlify/blob/4b9d6d2bd1466dc3e503b1220a20abf8f554133c/src/render.ts#L13)

```ts
const typeSymbol1 = Symbol('GraphQL Type'); // instead of this
const typeSymbol2 = Symbol.for('GraphQL Type'); // this
```

Then, I could make my own type producer

```ts
const enumerable = <T>(e: T): ValuesOf<T> => {
  return { [Symbol.for('GraphQL Type')]: GraphQLType.SCALAR, meta: { isEnum: true } } as any;
}
```

### Alternative 3

Alternative 2 seems that it's already possible since `typeSymbol` is exported. Another alternative I can think of is to include `typeSymbol` in the main export.

### Alternative 4 (no changes required)

I could always hack it on my end by producing a dummy type, and then using `Object.getOwnPropertySymbols()` to get `typeSymbol`.

```ts
const typeSymbol = Object.getOwnPropertySymbols(types.custom<any>())[0];
```
---

>Otherwise, I guess you are doing over engineering. If you are using graphql-code-generator then just only use it and forget about typed-graphqlify because that's much simpler to use for complex use-case.

That's what I was doing before I found `typed-graphqlify`. I ran into the same problem described in the [Motivation](https://github.com/acro5piano/typed-graphqlify#motivation).

The problem with using only `graphql-code-generator` is that you must check every field at runtime. After all, there is no typeable relationship between a query as a string and the data it produces. I kept having to write type assertion functions and type definitions. There was a lot of repeated code and it was very easy to make mistakes.

**Comment by acro5piano** [comment]
Okay, thank you for sharing it. I'm glad that you solved the issue.

Also, please consider converting your $gql into an standalone OSS library (or send a PR for it to typed-graphqlify) for others! The syntax is quite fluent and I'm curious how it works.

**Comment by jaredjj3** [comment]
Thanks for expressing interest. I'm currently in the middle of a refactor, so I will see how the solution evolves and consider making a standalone library after I fully understand what the limitations are.
